### ISSUE TITLE
training on other dataset

### ISSUE THREAD

**Issue by Jinho-P** [nianticlabs/monodepth2#183]
Hi !!
Thanks for sharing of awesome research.
I have some question about training monodepth2 using other dataset

1. in monodepth2 code, there is self.K intrinsic matrix normalized by image resolution(width, height) . 
When I use custom data (original_intrinsic, original_width, original_height), I resize image from (original_width, original_height) to (resized_width, resized_height). 
Now, I try to make custom self.K using original_intrinsic and (resized_width, resized_height).
I think custom self.K should be below
```
self.k[0,:] = original_intrinsic[0,:]/resized_width  # not original_width
self.k[1,:] = original_intrinsic[1,:]/resized_height # not resized_height
self.k[2,:] = [0,0,1,0]
self.k[3,:] = [0,0,0,1]
```
Is it correct?
And Is it better to use 0.5 fixed value than using actual normalized image center?

2. It is Necessary to make custom split txt file when using custom dataset
And I confirm that default split train_files and val_files contain shuffled data information.
The lines in split train_files and val_files should be shuffled?
If not shuffled, is there any a bad influence for training or performance?

3. When I try to train monodepth2 on A2D2 dataset from audi, 
I found the depth map estimation quality of training on A2D2 is worse than result on kitti dataset (default monodepth2 training result). 
(evaluation metric a1, a2, a3 is very low and result depth map looks like blurred)
Are there more things which I should change when using custom dataset?

**Comment by mrharicot** [comment]
Hi,
1. It should be normalized by the original size, as we are making it relative.
```python
self.k[0,:] = original_intrinsic[0,:] / original_width  # not resized_width
self.k[1,:] = original_intrinsic[1,:] / resized_height
```
It is not better to use 0.5 for the principal point, it was meant as a simplfication. If the pp isn't in the middle the intrinsics matrix needs to be modified when applying horizontal flips, which we currently do not do.
2. You can modify the datasets any way you want. We use txt splits as KITTI has multiple test splits and corresponding training splits. The lines don't need to be shuffled as the dataloader does this for you already. In general though, input data should be fed randomly.
3. Other than the intrinsics you shouldn't have to modify anything. If you implemented it like you said it is likely the cause of the bad results.

I hope this helps!

**Comment by Jinho-P** [comment]
Thanks for your reply !

I have one more question about previous question 1.
When I try to resize my training image, Is it necessary to keep width/height ratio of original image?
If I change image size with changed width/height ratio of original image, in this case, 
additional factor (ratio likes 'resized_width/original_width' & 'resized_height/original_height') should
be multiplied to self.k[0,:] and self.k[1,:] ?
(I resize a2d2 image from [1920, 1208] to [960, 608], 608 is 604 + 4 for multiple of 32)

In conclusion
case.1 :using normalized intrinsic matrix if train image was resized with same width/height ratio of original image.
```
self.k[0,:] = original_intrinsic[0,:] / original_width  # not resized_width
self.k[1,:] = original_intrinsic[1,:] / resized_height
```
case.2 : :using scaled normalized intrinsic matrix if train image was resized with different width/height ratio of original image.
```
self.k[0,:] = original_intrinsic[0,:] / original_width   * (resized_width/original_width)
self.k[1,:] = original_intrinsic[1,:] / resized_height  * (resized_height/original_height)
```

**Comment by mrharicot** [comment]
Hi,
1. You don't need to have the exact same ratio. The code takes care of scaling the instrincs to the new size later in the code, all you need is to normalize it by the original size. 
2. Automasking should ignore these sequences. If you see that they cause issues you might have to remove them.

Also [960, 608] is very large, it will take a while to train.

**Comment by Jinho-P** [comment]
1. I'm slightly confused about 'image size' because I don't think I clearly distinguish terms.
I means
- original image size : raw png image (original_width, height) (And normalized intrinsic matrix is calculated by (width, height) of original image)
- resize image size : resized png image (resized_width, height) -> this is used for train/valid dataset (i.e. in the case of kitti dataset, it is [1242, 375])
- input image size : image size used of network's input. It is determined at options '--width' and '--height'

In the cases I asked you above, I sets input image size to [640, 192] (default value of monodepth2) and resize image size to [960, 608] for used png image. 
In this situation, I wonder if what you have described above is still valid.
Also I think that input image size is still [640, 192], training time is not too large.

2. You mean that stationary image sequences are removed manually by human? 
I understand automasking can be used for ignore stationary image sequences from training time.
But is it my misunderstanding and for good performance, I should remove stationary image sequences are removed manually, right?

I'm sorry to asking similar question many times but I want to understand perfectly to this paper and working process.
Thank you for replying to me every time!!!

**Comment by mrharicot** [comment]
Hi,
1. You don't need to use the same input resolution as for KITTI, even if using a pretrained model. We used [640, 192]  to maintain the aspect ratio as much as possible but if you use another dataset you can use any resolution, as long as the size is a multiple of 32.
2. The automasking should ignore these stationary images so you shouldn't have to worry about it. But usually these sequences were removed "by hand" by computing optical flow and removing sequences with too little flow. See SfMLearner for details.

**Comment by Jinho-P** [comment]
Hi, @mrharicot 
I need additional advice for fixing problems...

As mention above, I trains monodepth2 with A2D2 dataset.
Raw image data has (1920, 1208) with intrinsic K [1687, 0, 965.4, 0, 1783, 684.4, 0, 0, 1] and it's aspect ratio is 1.58.

And I resize that image to size (352, 224) for having similar aspect ratio of raw image.
Then I trains monodepth2 with resized image (352, 224) with option '--width 352 --height 224' and modify self.K (in kitti_dataset.py) to [1687/1920, 0, 0.5, 0, 1783/1208, 0.5, 0, 0, 1]
(normalize intrinsic with original resolution and other training setting things are not changed)

Additional, I use frame_ids option with [0, -5, 5] because iamge change between adjacent frame is very small. So I modify that to similar with kitti data. Finally, I got 9804 train data and 1087 validation data. (batch_size = 12)
But I have bad result for accuracy, loss terms and depth quality also...

[accuracy, loss term]
![Screenshot from 2020-05-20 11-03-00](https://user-images.githubusercontent.com/41759110/82397434-6d1d7480-9a8b-11ea-94d4-077bbd40bba8.png)

[disparity map]
![Screenshot from 2020-05-20 11-04-30](https://user-images.githubusercontent.com/41759110/82397455-74448280-9a8b-11ea-8498-e57045eff74a.png)

[color image]
![Screenshot from 2020-05-20 11-05-16](https://user-images.githubusercontent.com/41759110/82397522-950cd800-9a8b-11ea-94fb-393127e438c7.png)

[predicted image]
![Screenshot from 2020-05-20 11-04-51](https://user-images.githubusercontent.com/41759110/82397536-9b02b900-9a8b-11ea-8387-b5f9ed06bc6a.png)

I analyzed it, but it was hard to find the cause of the performance drop.
Could you give me some advice?

**Comment by mrharicot** [comment]
Hi,
These depthmaps look pretty good to me!
The input resolution is pretty small, you could use something larger and experiment with other values than [-5, 5].
Finally 10k training samples is very very small, I would not expect a model to generalize very well with that little training data.

**Comment by Jinho-P** [comment]
Thanks for your reply!!

The result depth map looks good at glance, but comparing with default result about kitti, my result is more blurred and there are many wrong depth estimation.

I analysis change of my accuracy and losses. 
In training for your code with kitti, accuracy is increases steadily without huge fluctuation and also losses is decrease. So I think there are something wrong settings in my code...

I will try to various test with some questions.
1. you said "The input resolution is pretty small". Is ' input resolution ' means image resolution used training dataset or option values about '--width --height ?

2. In your training time, are there any similar cases that changing of accuracy and losses with huge fluctuation

Always appreciate!!

**Comment by mrharicot** [comment]
1. Yes I would re-export the original images to a higher resolution, say 512x320 or higher and change the options to match.
2. Your loss curves look indeed quite noisy, the depth errors should definitely go down. It is hard for me to assess whether this is normal as I have never used this dataset before but at least errors should go down more visibly and less noisy.
I would recommend to use a lot more training data.

**Comment by Jinho-P** [comment]
Thank you. 
I will try again with another setting.
If I has another questions, I will re-open this issue!!

**Comment by CaptainEven** [comment]
@mrharicot May i ask why the first row divided by original_width and the second row divided by resized height other than original_height?
```
self.k[0,:] = original_intrinsic[0,:] / original_width  # not resized_width
self.k[1,:] = original_intrinsic[1,:] / resized_height
```
